### PR TITLE
Alerting: Provisioning API - Alert rules

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
This PR is part of #36153 it introduces the API to create, update and delete alert rules.

### CODE REVIEW

File: pkg/services/ngalert/store/alert_rule.go [hunk 1]
```diff
@@ -162,15 +180,13 @@ func (st DBstore) InsertAlertRules(ctx context.Context, rules []ngmodels.AlertRu
...
-
 		return nil
 	})
 }
 
-// UpdateAlertRules is a handler for creating/updating alert rules.
+// UpdateAlertRules is a handler for updating alert rules.
 func (st DBstore) UpdateAlertRules(ctx context.Context, rules []UpdateRule) error {
 	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		newRules := make([]ngmodels.AlertRule, 0, len(rules))
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/store/alert_rule.go:173]: This was never used.

File: pkg/services/ngalert/store/alert_rule.go [hunk 2]
```diff
@@ -162,15 +180,13 @@ func (st DBstore) InsertAlertRules(ctx context.Context, rules []ngmodels.AlertRu
 				return fmt.Errorf("failed to create new rule versions: %w", err)
 			}
 		}
-
 		return nil
 	})
 }
 
-// UpdateAlertRules is a handler for creating/updating alert rules.
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/store/alert_rule.go:170]: Creating has never worked.

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	return *query.Result, provenance, nil
+}
+
+func (service *AlertRuleService) CreateAlertRule(ctx context.Context, rule models.AlertRule, provenance models.Provenance) (models.AlertRule, error) {
+	if rule.UID == "" {
+		rule.UID = util.GenerateShortUID()
+	}
+	interval, err := service.ruleStore.GetRuleGroupInterval(ctx, rule.OrgID, rule.NamespaceUID, rule.RuleGroup)
+	// if the alert group does not exists we just use the default interval
```
Comment (alexweav) [inline pkg/services/ngalert/provisioning/alert_rules.go:58]: Is it possible to create a rule with no group?

Should this create the group if it does not exist? If `UpdateRuleGroup` below behaves like an upsert, this probably works fine, but it's not clear from the name of that method.

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	// check that provenance is not changed in a invalid way
+	storedProvenance, err := service.provenanceStore.GetProvenance(ctx, rule, rule.OrgID)
+	if err != nil {
+		return err
+	}
+	if storedProvenance != provenance && storedProvenance != models.ProvenanceNone {
+		return fmt.Errorf("cannot delete with provided provenance '%s', needs '%s'", provenance, storedProvenance)
+	}
+	return service.xact.InTransaction(ctx, func(ctx context.Context) error {
+		err := service.ruleStore.DeleteAlertRulesByUID(ctx, orgID, ruleUID)
```
Comment (alexweav) [inline pkg/services/ngalert/provisioning/alert_rules.go:141]: If we delete the last rule in a group, should that also delete the group? (Maybe a thought for a future PR)

**Review by alexweav — APPROVED** [review]

**Comment by alexweav** [comment]
LGTM. My comments were generally small nitpicks and questions.

File: pkg/services/ngalert/api/api_provisioning.go [hunk 5]
```diff
@@ -223,6 +232,54 @@ func (srv *ProvisioningSrv) RouteDeleteMuteTiming(c *models.ReqContext) response
 	return response.JSON(http.StatusNoContent, nil)
 }
 
+func (srv *ProvisioningSrv) RouteRouteGetAlertRule(c *models.ReqContext) response.Response {
+	uid := web.Params(c.Req)[":UID"]
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/api_provisioning.go:236]: Changed!

File: pkg/services/ngalert/api/api_provisioning.go [hunk 6]
```diff
@@ -223,6 +232,54 @@ func (srv *ProvisioningSrv) RouteDeleteMuteTiming(c *models.ReqContext) response
...
+	return response.JSON(http.StatusOK, ar)
+}
+
+func (srv *ProvisioningSrv) RouteDeleteAlertRule(c *models.ReqContext) response.Response {
+	uid := web.Params(c.Req)[":UID"]
+	err := srv.alertRules.DeleteAlertRule(c.Req.Context(), c.OrgId, uid, alerting_models.ProvenanceAPI)
+	if err != nil {
+		return ErrResp(http.StatusInternalServerError, err, "")
+	}
+	return response.JSON(http.StatusOK, "")
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/api_provisioning.go:270]: Changed!

File: pkg/services/ngalert/api/api_provisioning.go [hunk 7]
```diff
@@ -223,6 +232,54 @@ func (srv *ProvisioningSrv) RouteDeleteMuteTiming(c *models.ReqContext) response
...
+
+func (srv *ProvisioningSrv) RoutePostAlertRule(c *models.ReqContext, ar apimodels.AlertRule) response.Response {
+	createdAlertRule, err := srv.alertRules.CreateAlertRule(c.Req.Context(), ar.UpstreamModel(), alerting_models.ProvenanceAPI)
+	if err != nil {
+		return ErrResp(http.StatusInternalServerError, err, "")
+	}
+	ar.ID = createdAlertRule.ID
+	ar.UID = createdAlertRule.UID
+	ar.Updated = createdAlertRule.Updated
+	return response.JSON(http.StatusOK, ar)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/api_provisioning.go:252]: Changed!

File: pkg/services/ngalert/api/generated_base_api_alertmanager.go [hunk 8]
```diff
@@ -4,7 +4,6 @@
  *
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/generated_base_api_alertmanager.go:4]: Mh not sure, I think making the makefile run gofmt is a good idea.

File: pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// swagger:route GET /api/provisioning/alert-rules/{UID} provisioning RouteGetAlertRule
+//
+// Get all the contact points.
+//
+//     Responses:
+//       200: AlertRule
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go:14]: It only returns one rule.

File: pkg/services/ngalert/provisioning/alert_rules.go (hunk 4 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/provisioning/alert_rules.go:141]: Talked about this offline, groups are part of the rules for now in the DB schema. So when the last rule of the group is deleted, it removes automatically the group as it's part of the rule record.

File: pkg/services/ngalert/api/authorization.go [hunk 10]
```diff
@@ -195,7 +196,11 @@ func (api *API) authorize(method, path string) web.Handler {
 		http.MethodDelete + "/api/provisioning/templates/{name}",
 		http.MethodPost + "/api/provisioning/mute-timings",
 		http.MethodPut + "/api/provisioning/mute-timings/{name}",
-		http.MethodDelete + "/api/provisioning/mute-timings/{name}":
+		http.MethodDelete + "/api/provisioning/mute-timings/{name}",
+		http.MethodPost + "/api/provisioning/alert-rules",
+		http.MethodPut + "/api/provisioning/alert-rules/{UID}",
+		http.MethodDelete + "/api/provisioning/alert-rules/{UID}",
+		http.MethodPut + "/api/provisioning/folder/{FolderUID}/rule-groups/{Group}":
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/authorization.go:203]: I think `group` is okay here, as a group has only a name it should be obvious.

**Pull request merged**